### PR TITLE
[Merged by Bors] - feat(GroupTheory/Coxeter): add auxiliary lemmas for alternating words and left inverse sequences

### PR DESCRIPTION
**Pull request by bolito2**
Add some auxiliary lemmas of alternating words and left inverse sequences of Coxeter groups that will be useful to prove the Strong Exchange Theorem (among other uses). They are mainly taken from pages 12-13 of [Combinatorics of Coxeter groups by Anders Bjorner , Francesco Brenti](https://link.springer.com/book/10.1007/3-540-27596-7).

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/GroupTheory/Coxeter/Basic.lean [hunk 1]
```diff
@@ -411,6 +411,87 @@ theorem length_alternatingWord (i i' : B) (m : ℕ) :
...
+      · rw [if_pos h_even]
+        rw[← even_add_two (n+k), ← Nat.add_assoc 2 n k] at h_even
+        rw [if_pos h_even]
+      · rw [if_neg h_even]
+        rw[← even_add_two (n+k), ← Nat.add_assoc 2 n k] at h_even
+        rw [if_neg h_even]
+
+lemma getElem_alternatingWord (i j : B) (p k : ℕ) (h : k < p) :
+    (alternatingWord i j p)[k]'(by simp; exact h) =  (if Even (p + k) then i else j) := by
+  simp[← getElem_alternatingWord_aux i j p ⟨k, (by simp[h])⟩]
```
[the remark this answers is not in the document]
Comment (bolito2) [reply Mathlib/GroupTheory/Coxeter/Basic.lean]: I managed to prove it directly 👍

File: Mathlib/GroupTheory/Coxeter/Basic.lean [hunk 2]
```diff
@@ -411,6 +411,78 @@ theorem length_alternatingWord (i i' : B) (m : ℕ) :
   · dsimp [alternatingWord]
   · simpa [alternatingWord] using ih i' i
 
+lemma getElem_alternatingWord (i j : B) (p k : ℕ) (hk : k < p) :
+    (alternatingWord i j p)[k]'(by simp; exact hk) =  (if Even (p + k) then i else j) := by
+  revert k
+  induction p with
+  | zero =>
+    intro k hk
+    simp[alternatingWord] at hk
```
Comment (jcommelin) [inline Mathlib/GroupTheory/Coxeter/Basic.lean:420]:
```suggestion
    simp [alternatingWord] at hk
```

File: Mathlib/GroupTheory/Coxeter/Basic.lean [hunk 3]
```diff
@@ -411,6 +411,78 @@ theorem length_alternatingWord (i i' : B) (m : ℕ) :
...
+    intro k hk
+    simp_rw [alternatingWord_succ' i j n]
+    match k with
+    | 0 =>
+      by_cases h2 : Even n
+      · simp [h2, (by simp[Even.add_one, h2]: ¬ Even (n + 1))]
+      · simp [h2, Odd.add_one (Nat.not_even_iff_odd.mp h2)]
+    | k + 1 =>
+      simp at hk h
+      simp[List.getElem_cons_succ, h k hk]
```
Comment (jcommelin) [inline Mathlib/GroupTheory/Coxeter/Basic.lean:431]:
```suggestion
      simp [List.getElem_cons_succ, h k hk]
```

File: Mathlib/GroupTheory/Coxeter/Basic.lean [hunk 4]
```diff
@@ -411,6 +411,78 @@ theorem length_alternatingWord (i i' : B) (m : ℕ) :
...
+      ring_nf
+      have even_add_two (m : ℕ) : Even (2 + m) ↔ Even m := by
+        simp[(Nat.even_sub (by omega: m ≤ 2 + m)).mp]
+      by_cases h_even : Even (n + k)
+      · rw [if_pos h_even]
+        rw[← even_add_two (n+k), ← Nat.add_assoc 2 n k] at h_even
```
Comment (jcommelin) [inline Mathlib/GroupTheory/Coxeter/Basic.lean:437]:
```suggestion
        rw [← even_add_two (n+k), ← Nat.add_assoc 2 n k] at h_even
```

File: Mathlib/GroupTheory/Coxeter/Basic.lean [hunk 5]
```diff
@@ -411,6 +411,78 @@ theorem length_alternatingWord (i i' : B) (m : ℕ) :
...
+        rw [if_pos h_even]
+      · rw [if_neg h_even]
+        rw[← even_add_two (n+k), ← Nat.add_assoc 2 n k] at h_even
+        rw [if_neg h_even]
+
+lemma getElem_alternatingWord_swapIndices (i j : B) (p k : ℕ) (h : k + 1 < p) :
+   (alternatingWord i j p)[k+1]'(by simp; exact h) =
+   (alternatingWord j i p)[k]'(by simp [h]; omega) := by
+  rw[ getElem_alternatingWord i j p (k+1) (by omega), getElem_alternatingWord j i p k (by omega)]
```
Comment (jcommelin) [inline Mathlib/GroupTheory/Coxeter/Basic.lean:446]:
```suggestion
  rw [getElem_alternatingWord i j p (k+1) (by omega), getElem_alternatingWord j i p k (by omega)]
```

**Review by jcommelin — COMMENTED** [review]
I've left a few stylistic comments, they also apply elsewhere.

I also have a more structural comment: for reasons of maintainability and robustness, we (roughly speaking) only want a bare `simp` at the end of a proof. In the middle of a proof, it should be a `simp only`.
See https://leanprover-community.github.io/glossary.html#non-terminal-simp for more details.

If you replace `simp` with `simp?`, it will suggest a robust `simp only` replacement.

This process is called "squeezing simp". So could you please "squeeze" the non-terminal simps in this PR?

**Review by jcommelin — COMMENTED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**